File: crates/symbi-a2ui/src/api/ws-types.ts
/**
 * WebSocket message types for the Coordinator Chat protocol.
 *
 * Mirrors the Rust `ws_types.rs` enums. The `type` field is the discriminant.
 */

// ---------------------------------------------------------------------------
// Client → Server
// ---------------------------------------------------------------------------

export interface ChatSend {
  type: 'ChatSend';
  /** Client-generated message id. */
  id: string;
  /** Natural-language content. */
  content: string;
}

export interface Ping {
  type: 'Ping';
}

export type ClientMessage = ChatSend | Ping;

// ---------------------------------------------------------------------------
// Server → Client
// ---------------------------------------------------------------------------

export interface ChatChunk {
  type: 'ChatChunk';
  request_id: string;
  content: string;
  done: boolean;
}

export interface ToolCallStarted {
  type: 'ToolCallStarted';
  request_id: string;
  call_id: string;
  tool_name: string;
  arguments: string;
}

export interface ToolCallResult {
  type: 'ToolCallResult';
  request_id: string;
  call_id: string;
  result: string;
  is_error: boolean;
}

export interface PolicyDecision {
  type: 'PolicyDecision';
  request_id: string;
  action: string;
  decision: string;
  reason: string;
}

export interface WsError {
  type: 'Error';
  request_id: string | null;
  code: string;
  message: string;
}

export interface Pong {
  type: 'Pong';
}

export type ServerMessage =
  | ChatChunk
  | ToolCallStarted
  | ToolCallResult
  | PolicyDecision
  | WsError
  | Pong;

File: crates/runtime/src/api/ws_types.rs
//! WebSocket message types for the Coordinator Chat protocol.
//!
//! Defines the client→server and server→client message envelopes used over
//! the `/ws/chat` WebSocket connection. Each `ChatSend` from the client
//! triggers a stream of `ServerMessage` variants correlated by `request_id`.

#[cfg(feature = "http-api")]
use serde::{Deserialize, Serialize};

/// Messages sent from the browser client to the server.
#[cfg(feature = "http-api")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientMessage {
    /// Send a chat message to the coordinator.
    ChatSend {
        /// Client-generated message id (for dedup / optimistic UI).
        id: String,
        /// Natural-language content.
        content: String,
    },
    /// Client ping (keepalive).
    Ping,
}

/// Messages sent from the server to the browser client.
#[cfg(feature = "http-api")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    /// A chunk of the assistant's streaming response.
    ChatChunk {
        /// Server-generated UUID correlating all events for one user message.
        request_id: String,
        /// Chunk content (may be empty on the final `done` message).
        content: String,
        /// `true` on the last chunk — signals the response is complete.
        done: bool,
    },
    /// A tool call has started executing.
    ToolCallStarted {
        request_id: String,
        call_id: String,
        tool_name: String,
        arguments: String,
    },
    /// A tool call has completed.
    ToolCallResult {
        request_id: String,
        call_id: String,
        result: String,
        is_error: bool,
    },
    /// A policy decision was made for an action.
    PolicyDecision {
        request_id: String,
        action: String,
        decision: String,
        reason: String,
    },
    /// An error occurred processing the request.
    Error {
        request_id: Option<String>,
        code: String,
        message: String,
    },
    /// Server pong (keepalive response).
    Pong,
}
